src/async_retry.rs

```diff
@@ -0,0 +1,212 @@
+    /// # }
+    /// ```
+    ///
+    /// or using `Box::pin` if params are required
+    ///
+    /// ```rust
+    /// # use backoff::{ExponentialBackoff, AsyncOperation, Error};
+    /// # use futures::future::BoxFuture;
+    /// # async fn main_task() {
+    /// let arg = "Async all the things!";
+    /// let mut f = || -> BoxFuture<Result<(), Error<&'static str>>> { Box::pin(async move {
+    ///     // Business logic...
+    ///     println!("{}", arg);
+    ///     // Give up.
+    ///     Err(Error::Permanent("error"))
+    /// })};
+    ///
+    /// let mut backoff = ExponentialBackoff::default();
+    /// let _ = f.retry(&mut backoff).await.err().unwrap();
+    /// # }
+    ///
+    /// # fn main() {
+    /// #    async_std::task::block_on(main_task());
+    /// # }
+    /// ```
+    async fn retry<B>(&mut self, backoff: &mut B) -> Result<T, Error<E>>
+    where
+        B: Backoff + Sync + Send,
+        T: 'async_trait,
+        E: 'async_trait,
+    {
+        let nop = |_, _| Box::pin(async {});
+        self.retry_notify(backoff, nop).await
+    }
+
+    /// Retries this operation according to the backoff policy.
+    /// Calls notify on failed attempts (in case of transient errors).
+    /// backoff is reset before it is used.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use backoff::{AsyncOperation, Error};
+    /// # use backoff::backoff::Stop;
+    /// # use std::time::Duration;
+    /// # use std::fmt::Display;
+    /// async fn notify<E>(err: E, dur: Duration)
+    /// where
+    ///     E: Display
+    /// {
+    ///     println!("Error happened at {:?}: {}", dur, err);
+    /// }
+    ///
+    /// async fn f() -> Result<(), Error<&'static str>> {
+    ///     // Business logic...
+    ///     Err(Error::Transient("error"))
+    /// }
+    ///
+    /// # async fn main_task() {
+    /// let mut backoff = Stop{};
+    /// let _ = f.retry_notify(&mut backoff, notify).await.err().unwrap();
+    /// # }
+    ///
+    /// # fn main() {
+    /// #    async_std::task::block_on(main_task());
+    /// # }
+    /// ```
+    async fn retry_notify<B, N>(&mut self, backoff: &mut B, mut notify: N) -> Result<T, Error<E>>
+    where
+        B: Backoff + Send,
+        T: 'async_trait,
+        E: 'async_trait,
+        N: AsyncNotify<E> + Send,
+    {
+        backoff.reset();
+
+        loop {
+            let err = match self.call_op().await {
+                Ok(v) => return Ok(v),
+                Err(err) => err,
+            };
+
+            let err = match err {
+                Error::Permanent(err) => return Err(Error::Permanent(err)),
+                Error::Transient(err) => err,
+            };
+
+            let next = match backoff.next_backoff() {
+                Some(next) => next,
+                None => return Err(Error::Transient(err)),
+            };
+
+            notify.notify(err, next).await;
+            async_std::task::sleep(next).await;
```

Comment:
And, also this way a long-wating `notify.notify(err, next).await` will delay  next retry tick much further than expected, because `async_std::task::sleep(next).await` is running only after notify is fully `.await`ed.

Reply:
Ah, so I need to spawn a task for that?